### ISSUE TITLE
False positive for Union[Literal[False], ClassName]

### ISSUE THREAD

**Issue by beeb** [python/mypy#8264]
* **Are you reporting a bug, or opening a feature request?**
Bug

* **Please insert below the code you are checking with mypy,
  or a mock-up repro if the source is private. We would appreciate
  if you try to simplify your case to a minimal repro.**

```python
from pathlib import Path
from typing_extensions import Literal  # support Python 3.7
a: Union[Literal[False], Path] = Path()
# mypy wrongly reports 'Item "bool" of "Union[Literal[False], Path]" has no attribute "exists"' but here we check that a is not False
if a and a.exists(): 
    print('the file exists')
```

* **What is the actual behavior/output?**

```Item "bool" of "Union[Literal[False], Path]" has no attribute "exists"```

* **What is the behavior/output you expect?**

No reported error

* **What are the versions of mypy and Python you are using?**
  **Do you see the same issue after installing mypy from Git master?**

mypy 0.760 - did not try installing from git master

* **What are the mypy flags you are using? (For example --strict-optional)**

```--ignore-missing-imports```

Hey there!

As you can see I found some unexpected behavior when trying to use the union of the literal "False" together with a class. It's a bit of a tricky one, as some variable can only be False or an instance of the class. I get that it's not a very common occurrence, since most people would just make it an optional, but nonetheless it's not the expected output from my point of view.

Thanks!

**Comment by JukkaL** [comment]
Yeah, this looks like a bug.

cc @Michael0x2a

**Comment by dosisod** [comment]
From: https://docs.python.org/3/library/typing.html#typing.cast

You may cast `a` to the desired type:

```py
from pathlib import Path
from typing_extensions import Literal
from typing import Union, cast

a: Union[Literal[False], Path] = Path()

if a and cast(Path, a).exists():
    print('the file exists')
```
mypy output:
```
Success: no issues found in 1 source file
```

**Comment by Michael0x2a** [comment]
It looks like this was fixed by https://github.com/python/mypy/pull/8368, which landed a few days ago.

**Comment by dosisod** [comment]
I have mypy installed from the master branch, yet these errors seem to still be present:

```
$ mypy --pretty test.py
test.py:7: error: Item "bool" of "Union[Literal[False], Path]" has no attribute "exists"
    if a and a.exists():
             ^
Found 1 error in 1 file (checked 1 source file)

$ mypy --version
mypy 0.770+dev.d6c2c01fd04043bff55222f1f43a9fe2e15c0ccf
```

**Comment by Michael0x2a** [comment]
@dosisod -- I don't seem to be able to repro. Running the following example against the same mypy checkout you're using results in no errors and a revealed type of `pathlib.Path` for me:

```python
from pathlib import Path
from typing_extensions import Literal
from typing import Union

a: Union[Literal[False], Path] = Path()
if a and a.exists(): 
    reveal_type(a)
    print('the file exists')
```

Maybe try clearing your mypy cache?

**Comment by dosisod** [comment]
This fixed my problem, thanks.
